op-node/rollup/derive/frame_queue_test.go

```diff
@@ -0,0 +1,154 @@
+package derive
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"log/slog"
+	"testing"
+
+	"github.com/ethereum-optimism/optimism/op-node/rollup"
+	"github.com/ethereum-optimism/optimism/op-node/rollup/derive/mocks"
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+	"github.com/ethereum-optimism/optimism/op-service/testlog"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+)
+
+func TestPruneFrameQueue(t *testing.T) {
+	for _, tt := range []struct {
+		desc string
+		fs   []testFrame
+		exp  []testFrame
+	}{
+		{
+			desc: "empty",
+			fs:   []testFrame{},
+			exp:  []testFrame{},
+		},
+		{
+			desc: "one",
+			fs:   []testFrame{"a:2:"},
+			exp:  []testFrame{"a:2:"},
+		},
+		{
+			desc: "one-last",
+			fs:   []testFrame{"a:2:!"},
+			exp:  []testFrame{"a:2:!"},
+		},
+		{
+			desc: "last-new",
+			fs:   []testFrame{"a:2:!", "b:0:"},
+			exp:  []testFrame{"a:2:!", "b:0:"},
+		},
+		{
+			desc: "last-ooo",
+			fs:   []testFrame{"a:2:!", "b:1:"},
+			exp:  []testFrame{"a:2:!"},
+		},
+		{
+			desc: "middle-lastooo",
+			fs:   []testFrame{"b:1:", "a:2:!"},
+			exp:  []testFrame{"b:1:"},
+		},
+		{
+			desc: "middle-first",
+			fs:   []testFrame{"b:1:", "a:0:"},
+			exp:  []testFrame{"a:0:"},
```

Comment:
This behaviour matches [the specs](https://specs.optimism.io/protocol/holocene/derivation.html?highlight=hol#frame-queue). But I'm curious about whether it would have been equally good to drop `a:0:` instead of `b:1:` in this case? Wouldn't that maintain the ordering that we want and also reduce the chance of a reorg?

Reply:
That would be possible in theory, but have the disadvantage that the DP could only ever make progress if that last closing frame really arrives. It's much better to be able to just implicitly throw away a half-open channel by a newly incoming first frame. This way, any partial channel submission doesn't have any negative impact from a batcher restart, and a batcher restart should be able to immediately fix any batcher tx inconsistencies from a past batcher instance. Plus, this rule also improves the optimistic sync start behavior, where if we find a first frame, we're guaranteed that any previous non-closed frame data would have been dropped and so couldn't influence the DP state.